Userland/Libraries/LibTextCodec/Encoder.cpp

```diff
@@ -0,0 +1,518 @@
+    //    in index ISO-2022-JP katakana.
+    if (item >= 0xFF61 && item <= 0xFF9F) {
+        item = *index_iso_2022_jp_katakana_code_point(item - 0xFF61);
+    }
+
+    // 10. Let pointer be the index pointer for code point in index jis0208.
+    auto pointer = code_point_jis0208_index(item);
+
+    // 11. If pointer is null, then:
+    if (!pointer.has_value()) {
+        // 1. If ISO-2022-JP encoder state is jis0208, then restore code point to ioQueue, set ISO-2022-JP encoder state to
+        //    ASCII, and return three bytes 0x1B 0x28 0x42.
+        if (state == State::jis0208) {
+            TRY(on_byte(0x1B));
+            TRY(on_byte(0x28));
+            TRY(on_byte(0x4A));
+            return process_item(item, State::ASCII, on_byte, on_error);
+        }
+
+        // 2. Return error with code point.
+        TRY(on_error(item));
+        return state;
+    }
+
+    // 12. If ISO-2022-JP encoder state is not jis0208, restore code point to ioQueue, set ISO-2022-JP encoder state to
+    //     jis0208, and return three bytes 0x1B 0x24 0x42.
+    if (state != State::jis0208) {
+        TRY(on_byte(0x1B));
+        TRY(on_byte(0x24));
+        TRY(on_byte(0x42));
+        return process_item(item, State::jis0208, on_byte, on_error);
+    }
+
+    // 13. Let lead be pointer / 94 + 0x21.
+    auto lead = *pointer / 94 + 0x21;
+
+    // 14. Let trail be pointer % 94 + 0x21.
+    auto trail = *pointer % 94 + 0x21;
+
+    // 15. Return two bytes whose values are lead and trail.
+    TRY(on_byte(static_cast<u8>(lead)));
+    TRY(on_byte(static_cast<u8>(trail)));
+    return state;
+}
+
+// https://encoding.spec.whatwg.org/#iso-2022-jp-encoder
+ErrorOr<void> ISO2022JPEncoder::process(Utf8View input, Function<ErrorOr<void>(u8)> on_byte, Function<ErrorOr<void>(u32)> on_error)
+{
+    // ISO-2022-JP’s encoder has an associated ISO-2022-JP encoder state which is ASCII, Roman, or jis0208 (initially ASCII).
+    auto state = State::ASCII;
+
+    for (u32 item : input) {
+        state = TRY(process_item(item, state, on_byte, on_error));
+    }
+
+    // 1. If code point is end-of-queue and ISO-2022-JP encoder state is not ASCII, set ISO-2022-JP
+    //    encoder state to ASCII, and return three bytes 0x1B 0x28 0x42.
+    if (state != State::ASCII) {
+        state = State::ASCII;
+        TRY(on_byte(0x1B));
+        TRY(on_byte(0x28));
+        TRY(on_byte(0x42));
+        return {};
+    }
+
+    // 2. If code point is end-of-queue and ISO-2022-JP encoder state is ASCII, return finished.
+    return {};
+}
+
+// https://encoding.spec.whatwg.org/#index-shift_jis-pointer
+static Optional<u32> index_shift_jis_pointer(u32 code_point)
+{
+    // 1. Let index be index jis0208 excluding all entries whose pointer is in the range 8272 to 8835, inclusive.
+    auto pointer = code_point_jis0208_index(code_point);
+    if (!pointer.has_value())
+        return {};
+    if (*pointer >= 8272 && *pointer <= 8835)
+        return {};
```

Comment:
I don't think this is correct. There is a note on this spec step that says:
> The [index jis0208](https://encoding.spec.whatwg.org/#index-jis0208) contains duplicate code points so the exclusion of these entries causes later code points to be used.

So, I think you need to modify `code_point_jis0208_index()` so that its for loop skips these indices, rather than returning nothing. 

This should fix the issues I'm seeing on: http://wpt.live/encoding/legacy-mb-japanese/shift_jis/sjis-encode-form-ms932.html, for example.

Reply:
You're exactly right, changing it to skip instead of ignoring these pointers fixes those tests.

Reply:
Nice! Please click the `[Resolve conversation]` button for feedback that you've dealt with. :^)